hazelcast/src/main/java/com/hazelcast/map/impl/operation/WriteBehindStateHolder.java

```diff
@@ -144,42 +126,60 @@ public void writeData(ObjectDataOutput out) throws IOException {
         MapServiceContext mapServiceContext = mapService.getMapServiceContext();
 
-        out.writeInt(delayedEntries.size());
-        for (Map.Entry<String, List<DelayedEntry>> entry : delayedEntries.entrySet()) {
-            out.writeUTF(entry.getKey());
-            List<DelayedEntry> delayedEntryList = entry.getValue();
-            out.writeInt(delayedEntryList.size());
-
-            for (DelayedEntry e : delayedEntryList) {
-                Data key = mapServiceContext.toData(e.getKey());
-                Data value = mapServiceContext.toData(e.getValue());
-                long expirationTime = e.getExpirationTime();
-
-                IOUtil.writeData(out, key);
-                IOUtil.writeData(out, value);
-                out.writeLong(expirationTime);
-                out.writeLong(e.getStoreTime());
-                out.writeInt(e.getPartitionId());
-                out.writeLong(e.getSequence());
-                UUIDSerializationUtil.writeUUID(out, e.getTxnId());
-            }
+        // serialize data in wbq
+        out.writeInt(stores.size());
+        for (WriteBehindStore store : stores) {
+            out.writeUTF(store.getMapName());
+            store.forEach(new WriteBehindQueueConsumer<DelayedEntry>() {
+
+                @Override
+                public void size(int size) {
+                    try {
+                        out.writeInt(size);
+                    } catch (Exception e) {
+                        throw rethrow(e);
+                    }
+                }
+
+                @Override
+                public void accept(DelayedEntry delayedEntry) {
+                    try {
+                        Data key = mapServiceContext.toData(delayedEntry.getKey());
+                        Data value = mapServiceContext.toData(delayedEntry.getValue());
+                        long expirationTime = delayedEntry.getExpirationTime();
+
+                        IOUtil.writeData(out, key);
+                        IOUtil.writeData(out, value);
+                        out.writeLong(expirationTime);
+                        out.writeLong(delayedEntry.getStoreTime());
+                        out.writeInt(delayedEntry.getPartitionId());
+                        out.writeLong(delayedEntry.getSequence());
+                        UUIDSerializationUtil.writeUUID(out, delayedEntry.getTxnId());
+                    } catch (Exception e) {
+                        throw rethrow(e);
+                    }
+                }
+            });
         }
 
-        out.writeInt(flushSequences.size());
-        for (Map.Entry<String, Queue<WriteBehindStore.Sequence>> entry : flushSequences.entrySet()) {
-            out.writeUTF(entry.getKey());
-            Queue<WriteBehindStore.Sequence> queue = entry.getValue();
-            out.writeInt(queue.size());
-            for (WriteBehindStore.Sequence sequence : queue) {
+        // serialize flush sequences
+        out.writeInt(stores.size());
```

Comment:
I guess we could not write the store size three times, but let's not change it now.